llvm/lib/Target/AIE/AIEWawRegRewriter.cpp

```diff
@@ -0,0 +1,381 @@
+  // writing into subregisters does not already trigger the register
+  // replacement. In this case $x0 would have already been replaced, even though
+  // there is no real WAW dependency.
+  //                                                                          \
+  // undef %0.sub_256_lo:mxa, _, _= VLDA_2D_dmw_lda_w _,_                     \
+  // % 0.sub_256_hi:vec512 = VLDA_dmw_lda_w_ag_idx_imm _,_                    \
+  // %7:mxm, %8:el = VMAX_LT_D8 %0, $x6                                       \
+  //
+  bool isLastDef(const MachineOperand &Def, const MachineInstr &DefInstr,
+                 const MachineBasicBlock &MBB);
+
+  // Block every sub- and super-register of a physical register, so that it is
+  // removed for future replacement strategies, i.e. block wl4, wh4, y2 if X4
+  // is used.
+  void addAliasRegs(BitVector &BlockedPhysRegs, const MCPhysReg &PhysReg);
+};
+
+MCPhysReg
+AIEWawRegRewriter::getAssignedPhysReg(const MachineOperand &MO) const {
+  Register Reg = MO.getReg();
+
+  if (Reg.isStack())
+    return MCRegister::NoRegister;
+
+  if (Reg.isVirtual())
+    Reg = VRM->getPhys(Reg);
+
+  return Reg;
+}
+
+bool AIEWawRegRewriter::runOnMachineFunction(MachineFunction &MF) {
+
+  SmallVector<const MachineBasicBlock *, 4> LoopMBBs =
+      AIELoopUtils::getSingleBlockLoopMBBs(MF);
+
+  if (LoopMBBs.empty())
+    return false;
+
+  this->MF = &MF;
+  MRI = &MF.getRegInfo();
+  TRI = static_cast<const AIEBaseRegisterInfo *>(MRI->getTargetRegisterInfo());
+  VRM = &getAnalysis<VirtRegMap>();
+  LRM = &getAnalysis<LiveRegMatrix>();
+  LIS = &getAnalysis<LiveIntervals>();
+  Modified = false;
+
+  LLVM_DEBUG(dbgs() << "*** WAW Loop Register Rewriting: " << MF.getName());
+  LLVM_DEBUG(dbgs() << " ***\n");
+
+  for (const MachineBasicBlock *MBB : LoopMBBs)
+    renameMBBPhysRegs(MBB);
+
+  return Modified;
+}
+
+// return a vector containing virtual registers, that have had a copy to define
+// the virtual register
+SmallVector<Register, 16> getVRegWithCopies(const MachineBasicBlock &MBB) {
+  SmallVector<Register, 16> VRegWithCopies;
+  for (const MachineInstr &MI : MBB) {
+    for (const MachineOperand Def : MI.defs()) {
+      Register Reg = Def.getReg();
+
+      if (!Reg.isVirtual())
+        continue;
+
+      if (MI.isCopyLike() && find(VRegWithCopies, Reg) == VRegWithCopies.end())
+        VRegWithCopies.push_back(Reg);
+    }
+  }
+
+  return VRegWithCopies;
+}
+
+bool AIEWawRegRewriter::renameMBBPhysRegs(const MachineBasicBlock *MBB) {
+  LLVM_DEBUG(dbgs() << "WAW Reg Renaming BasicBlock "; MBB->dump();
+             dbgs() << "\n");
+  // Add the used physical registers one machine instruction at at time.
+  // This vector is used to determine, if a physical register has already been
+  // defined in the machine basic block.
+  BitVector UsedPhysRegs(TRI->getNumRegs());
+
+  // record the register replacements for debugging purposes.
+  SmallVector<std::pair<MCPhysReg, MCPhysReg>, 8> ReplacementRegs;
```

Comment:
Nit: Could we dump info about the renaming immediately? I'd rather avoid variables that are only useful for debugging

Reply:
I already dump the renaming information, it was just a nice way of debugging large functions, where the unsuccessful replacements were crowding out the successful ones.

I removed it.